### PR TITLE
#26395 Disabling possibly unnecessary prefetching during GroupIntoBatches by using an experimental flag

### LINKED ISSUE #26395: [Bug]: Possibly unnecessary prefetch during GroupIntoBatches
### What happened?

As I was inspecting why the Dataflow "Total streaming data processed" metric was so much higher (2-4× depending on the pipeline) than the actual data that was being processed as it only has one shuffle/gbk/gib/etc transform I stumbled upon something that I couldn't justify being there:
https://github.com/apache/beam/blob/f549fd33abdc672143ccbe3f0f66104995d30fe6/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java#L566-L569

I might be missing some context here, but I'm not even sure what benefit this provides for any generic `GroupIntoBatches` transform, but it seems to be especially useless if called on a `GroupIntoBatches.WithShardedKey` (when every data the batch contains comes from the same worker already).

I would appreciate any insight before I submit a PR to change this.

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [X] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by nbali**
We haven't reached a verdict on the proper way to implement #26395, so I created a PR for all the alternatives. The chosen one should be reviewed+merged, the rest should be closed. (alternative PR: https://github.com/apache/beam/pull/26619)

closes #26395

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - ~[ ] Update `CHANGES.md` with noteworthy changes.~
 - ~[ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).~

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nbali** [comment]
seemingly irrelevant failure during `Java Tests / Java Wordcount Direct Runner (ubuntu-latest) (pull_request)`

**Comment by nbali** [comment]
Run Java_Pulsar_IO_Direct PreCommit

**Comment by nbali** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java [hunk 1]
```diff
@@ -110,6 +112,10 @@
 public class GroupIntoBatches<K, InputT>
     extends PTransform<PCollection<KV<K, InputT>>, PCollection<KV<K, Iterable<InputT>>>> {
 
+  /** Experiment to "avoid possibly unnecessary prefetching". */
+  public static final String AVOID_POSSIBLY_UNNECESSARY_PREFETCHING =
+      "avoid_possibly_unnecessary_prefetching";
```
Comment (Abacn) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java:117]: consider a descriptive naming: disable_groupintobatches_preferch

**Review by Abacn — COMMENTED** [review]
LGTM Thanks. Just one naming comment. I would vote to go through with this approach at this moment as it does not change the current setting; or can be one step further to make it opt-out, but not completely remove it at this moment

**Comment by Abacn** [comment]
Tested and found this might not be the cause of data processed being much larger than the actual data size: https://github.com/apache/beam/issues/26395#issuecomment-1546138402

**Comment by tvalentyn** [comment]
waiting on author

**Comment by nbali** [comment]
FYI I'm trying to make time for this this week, but it's most likely it won't happen. If there is a way to make the reminders snooze for 1-2 weeks feel free to do so.

**Comment by tvalentyn** [comment]
waiting on author

**Comment by hpvd** [comment]
added this to [Parent issue] Support for Apache Pulsar #31078

**Comment by nbali** [comment]
@hpvd I'm pretty sure it's not related to Apache Pulsar

**Comment by hpvd** [comment]
> @hpvd I'm pretty sure it's not related to Apache Pulsar

@nbali thanks, I have removed it.

**Pull request closed without merging**